refactor apply_w8a8_block_fp8_linear in fp

ChangyiYang opened the pull request:
<!-- Thank you for your contribution! We appreciate it. The following guidelines will help improve your pull request and facilitate feedback. If anything is unclear, don't hesitate to submit your pull request and ask the maintainers for assistance. -->

See also: #4353



## Motivation

refactor apply_w8a8_block_fp8_linear to make the logic more clear and more meaningful

## Modifications

only refactor apply_w8a8_block_fp8_linear in fb8_utils

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

ChangyiYang committed:
refactor apply_w8a8_block_fp8_linear

Fridge003 commented on python/sglang/srt/layers/quantization/fp8.py:
```diff
@@ -417,7 +419,7 @@ def apply(
 
         if self.block_quant:
-            return apply_w8a8_block_fp8_linear(
+            return self.w8a8_block_fp8_linear(
```
I feel there is no need to rename `apply_w8a8_block_fp8_linear`. It can be used as before, with the dispatching logic rewritten.

ChangyiYang replied:
@Alcanderian Could you please have a look? On my end, I also feel that the method name could be remain unchanged. Calling it dispatch might give the impression that it only handles dispatching, rather than performing any computation.

Alcanderian replied:
> I feel there is no need to rename `apply_w8a8_block_fp8_linear`. It can be used as before, with the dispatching logic rewritten.

Do you mean recover it to `self.apply_w8a8_block_fp8_linear`? That's acceptable

Alcanderian commented on python/sglang/srt/layers/quantization/fp8_utils.py:
```diff
@@ -137,46 +137,85 @@ def apply_w8a8_block_fp8_linear(
-        )
-        output = fp8_blockwise_scaled_mm(
-            q_input, weight.T, x_scale, weight_scale.T, out_dtype=input.dtype
+        return cutlass_w8a8_block_fp8_linear(
```
This logic seems not right, my expection is just call `return cutlass_w8a8_block_fp8_linear` here. That means we are assgining `cutlass_w8a8_block_fp8_linear` to `self.w8a8_block_fp8_linear`, and making the calling phase at `return self.w8a8_block_fp8_linear(...)` do not need to filter implementations. @Fridge003 @ChangyiYang How do you think about that?

ChangyiYang replied:
@Alcanderian Hi Alcanderian! Correct me if I'm wrong, but as I understand it, we need to know the shape of the layer.weight to determine whether we can use Cutlass:

```
shape_supported_by_cutlass = (
    weight.shape[0] % 128 == 0 and weight.shape[1] % 128 == 0
)
```


Since this shape information isn't available during `__init__`, I think we can't fully resolve the dispatch logic at that point. One potential workaround could be to dynamically check for Cutlass compatibility at runtime (based on the actual weight shape), while falling back to a statically chosen default implementation for the other cases during`__init__`.

ChangyiYang replied:
Got it. So would it be suitable if I just dynamically check for Cutlass compatibility at runtime and fallback to trition implementation if shape isn't right?

Alcanderian replied:
I see, we may have to check `weight.shape[0] % 128 == 0 and weight.shape[1] % 128 == 0` in both cutlass and deepgemm(aslo `C.dtype == torch.bfloat16` for deepgemm and extract deepgemm from w8a8_block_fp8_matmul, [ref](https://github.com/sgl-project/sglang/blob/main/python/sglang/srt/layers/quantization/fp8_kernel.py#L784)) and fallback to triton if false.

The point is avoiding more and more filtering overhead from introducing implementations and vendors.

ChangyiYang replied:
Ok will do that

ChangyiYang committed:
refactoring the dispatching logic to avoid filtering overhead

ChangyiYang commented:
Hi! I have create a new commit. Here are some bullet points
* Refactored dispatch logic to return only the selected matmul implementation function at initialization time.
* For implementations requiring runtime validation (e.g., shape or dtype constraints), added fallback to Triton when conditions are not met.
* Moved runtime condition checks ahead of computation to avoid redundant operations and unnecessary overhead.
* Extracted w8a8_block_fp8_matmul_deepgemm from w8a8_block_fp8_matmul to enable direct usage of DeepGEMM kernels when applicable.
* Retained the function w8a8_block_fp8_matmul as the unified entry point for matrix multiplication, to avoid requiring users to manually check whether DeepGEMM is available each time they use it.

ChangyiYang commented on python/sglang/srt/layers/quantization/fp8_kernel.py:
```diff
@@ -764,77 +816,62 @@ def w8a8_block_fp8_matmul(
-    assert A.shape[:-1] == As.shape[:-1] and A.is_contiguous()
-    assert triton.cdiv(A.shape[-1], block_k) == As.shape[-1]
-    M = A.numel() // A.shape[-1]
+    if output_dtype == torch.bfloat16 and _ENABLE_JIT_DEEPGEMM:
```
Yes for example in [this test](https://github.com/sgl-project/sglang/blob/e235be16fe720f62aed1ec1dd3dbf7b9dfcf2107/python/sglang/test/test_block_fp8.py#L14). 

An alternatvie way is I can remove this branch in `w8a8_block_fp8_matmul`, rename it to  `w8a8_block_fp8_matmul_triton`, Then just call corrsponding matmul funtion in that test.

Fridge003 replied:
@Alcanderian May you have a look

Fridge003 replied:
@Alcanderian  I just checked the kernel, cutlass w8a8 supports both fp16 and bf16.

Alcanderian replied:
> @Alcanderian so should I do the branch remove stuff I talked about now? Any more modification needed for this PR?
> 
> > Yes for example in [this test](https://github.com/sgl-project/sglang/blob/e235be16fe720f62aed1ec1dd3dbf7b9dfcf2107/python/sglang/test/test_block_fp8.py#L14).
> > An alternatvie way is I can remove this branch in `w8a8_block_fp8_matmul`, rename it to `w8a8_block_fp8_matmul_triton`, Then just call corrsponding matmul funtion in that test.
> > Also in benchmark file [here](https://github.com/sgl-project/sglang/blob/e235be16fe720f62aed1ec1dd3dbf7b9dfcf2107/benchmark/kernels/deepseek/benchmark_deepgemm_fp8_gemm.py#L10) , but I think for this one I can directly call w8a8_block_fp8_matmul_deepgemm

I found we have check datatype at [here](https://github.com/sgl-project/sglang/pull/6545/files#diff-a6a536f2f2f27489d809052e552694dd151010a75d09adeeda3a02cb78e4399fR189). So deep_gemm will be fallback to triton in the caller, we can separate out `w8a8_block_fp8_matmul_triton` 🎉

ChangyiYang committed:
Modify comments in fp8_kernel.py

Co-authored-by: Xiaoyu Zhang <35585791+BBuf@users.noreply.github.com>

ChangyiYang committed:
create w8a8_block_fp8_matmul_triton, leave w8a8_block_fp8_matmul as a universal entry for testing purpose

ChangyiYang commented:
Hi ! Here are some modification for this commit 
* split out `w8a8_block_fp8_matmul_triton`
* retain `w8a8_block_fp8_matmul` only for testing purpose
* for benchmarks that use deepgemm, directly call w8a8_block_fp8_matmul_deepgemm

Feel free to tell me if there anything needs further modification!

Alcanderian commented on sgl-kernel/benchmark/bench_fp8_blockwise_gemm.py:
```diff
@@ -9,7 +9,9 @@
 
-from sglang.srt.layers.quantization.fp8_kernel import w8a8_block_fp8_matmul
+from sglang.srt.layers.quantization.fp8_kernel import (
+    w8a8_block_fp8_matmul_gemm as w8a8_block_fp8_matmul,
```
Is `w8a8_block_fp8_matmul_gemm` typo?

ChangyiYang replied:
sry fix that!

ChangyiYang committed:
fix typo

ChangyiYang commented:
Hi! Typo fixed. Feel free to tell me if there is any more issue!

zhyncs commented:
cc @HaiShaw

ChangyiYang commented:
@Alcanderian hi, I click update branch to merge from `main` and seems needs to be approved again. Can you kindly approve again?

Alcanderian commented on sgl-kernel/benchmark/bench_fp8_blockwise_gemm.py:
```diff
@@ -9,7 +9,9 @@
 
-from sglang.srt.layers.quantization.fp8_kernel import w8a8_block_fp8_matmul
+from sglang.srt.layers.quantization.fp8_kernel import (
+    w8a8_block_fp8_matmul_deepgemm as w8a8_block_fp8_matmul,
```
```suggestion
    w8a8_block_fp8_matmul_triton as w8a8_block_fp8_matmul,
```
Ref: [this line](https://github.com/sgl-project/sglang/pull/6545/files#diff-30c0296c13657392b6438f93c0ac211d99b4f18e3aa87a5db4d893b3d667aa40R135)

ChangyiYang committed:
Update kernel function ref in bench_fp8_blockwise_gemm.py

Co-authored-by: JieXin Liang <Alcanderian@users.noreply.github.com>

Alcanderian committed:
Merge branch 'main' into refactor_apply_w8a8_block_fp8_linear

Alcanderian commented:
@ChangyiYang We have to fix the CI
https://github.com/sgl-project/sglang/actions/runs/15267675566/job/42936151803?pr=6545#step:4:325
https://github.com/sgl-project/sglang/actions/runs/15267675566/job/42936151795?pr=6545#step:4:339
https://github.com/sgl-project/sglang/actions/runs/15267675566/job/42936151794?pr=6545#step:4:3685
https://github.com/sgl-project/sglang/actions/runs/15267675566/job/42936151793?pr=6545#step:4:1352

ChangyiYang committed:
fix referenced before assignment error

ChangyiYang commented:
@Alcanderian thanks for pointing that out! fix that now. Can you run CI again?

ChangyiYang committed:
fix bug that output_dtype is not passed correctly

ChangyiYang committed:
fix bug that output_dtype is not passed correctly

Alcanderian commented:
> Hi! I think I have fixed 2 minor bugs and the CI seems be correct ( the failing one is not actually failing and 9 pipelines are blocking). Feel free to tell me if any more adjustment needed before merge! Thanks you all guys for the help :)

The failing remained could be ignore now

zhyncs committed:
Merge branch 'main' into refactor_apply_w8a8_block_fp8_linear

whybeyoung committed:
[PD Perf] replace Queue to FastQueue (#6649)

Signed-off-by: Shangming Cai <caishangming@linux.alibaba.com>
Co-authored-by: Shangming Cai <caishangming@linux.alibaba.com>

ShangmingCai committed:
[Bugfix] Fix slice operation when chunk size mismatch (#6697)

Signed-off-by: Shangming Cai <caishangming@linux.alibaba.com>

ShangmingCai committed:
[Bugfix] Fix ChatCompletion endpoint of mini_lb when stream is set (#6703)

Signed-off-by: Shangming Cai <caishangming@linux.alibaba.com>

ShangmingCai committed:
[CI] Fix setup of disaggregation with different tp (#6706)

Signed-off-by: Shangming Cai <caishangming@linux.alibaba.com>

Hongbosherlock committed:
[PD] Remove Unnecessary Exception Handling for FastQueue.get() (#6712)

fzyzcjy committed:
Fuse routed_scaling_factor in DeepSeek (#6710)

fzyzcjy committed:
Overlap two kernels in DeepSeek with communication (#6711)

fzyzcjy committed:
Minor refactor two-batch overlap (#6682)

fzyzcjy committed:
Speed up when having padding tokens two-batch overlap (#6668)

Co-authored-by: Cheng Wan <54331508+ch-wan@users.noreply.github.com>

Fridge003 committed:
[Feature] Support Flashinfer fp8 blockwise GEMM kernel on Blackwell (#6479)

Fridge003 commented:
Hi @ChangyiYang , please fix the conflict caused by new flashinfer gemm method in #6479

ChangyiYang committed:
create new dispatching function flashinfer_gemm_w8a8_block_fp8_linear

ChangyiYang committed:
Merge branch 'main' into refactor_apply_w8a8_block_fp8_linear

ChangyiYang commented:
Hi! Conflict solved! Add a new method `flashinfer_gemm_w8a8_block_fp8_linear`

ChangyiYang committed:
Merge branch 'main' into refactor_apply_w8a8_block_fp8_linear

ChangyiYang committed:
Merge branch 'main' into refactor_apply_w8a8_block_fp8_linear

Fridge003 reviewed (approved):
LGTM

zhyncs committed:
Merge branch 'main' into refactor_apply_w8a8_block_fp8_linear
